keel-core/src/test/kotlin/com/netflix/spinnaker/keel/actuation/EnvironmentPromotionCheckerTests.kt

```diff
@@ -319,7 +357,129 @@ internal class EnvironmentPromotionCheckerTests : JUnit5Minutests {
 
         test("no artifact is registered") {
           verify(exactly = 0) {
-            artifactRepository.approveVersionFor(deliveryConfig, artifact, "1.0", environment.name)
+            repository.approveVersionFor(deliveryConfig, artifact, any(), environment.name)
+          }
+        }
+      }
+
+      context("the environment has a stateful constraint and a version cannot be found") {
+        deriveFixture {
+          copy(environment = Environment(
+            name = "staging",
+            constraints = setOf(ManualJudgementConstraint())
+          ))
+        }
+
+        before {
+          // TODO: sucks that this is necessary but when using deriveFixture you get a different mockk
+          every { repository.artifactVersions(artifact) } returns listOf("2.0", "1.2", "1.1")
+
+          every { repository.vetoedEnvironmentVersions(any()) } returns emptyList()
+
+          every { repository.pinnedEnvironments(any()) } returns emptyList()
+
+          every { repository.vetoedEnvironmentVersions(any()) } returns emptyList()
+
+          every { repository.pendingConstraintVersionsFor(any(), any()) } returns emptyList()
+
+          every { repository.getQueuedConstraintApprovals(any(), any()) } returns emptySet()
+
+          every { statefulEvaluator.canPromote(artifact, "2.0", deliveryConfig, environment) } returns false
+
+          every {
+            repository.constraintStateFor("my-manifest", "staging", "2.0")
+          } returns listOf(pendingManualJudgement)
+
+          runBlocking { subject.checkEnvironments(deliveryConfig) }
+        }
+
+        test("stateful constraints are only evaluated for the most recent version") {
+          verify(exactly = 1) {
+            statefulEvaluator.canPromote(artifact, "2.0", deliveryConfig, environment)
+          }
+          verify(exactly = 0) {
+            statefulEvaluator.canPromote(artifact, "1.2", deliveryConfig, environment)
+            statefulEvaluator.canPromote(artifact, "1.1", deliveryConfig, environment)
+          }
+        }
+
+        test("no artifact is approved") {
+          verify(exactly = 0) {
+            repository.approveVersionFor(deliveryConfig, artifact, any(), environment.name)
+          }
+        }
+
+        test("no exception is thrown") {
+          expectCatching {
+            subject.checkEnvironments(deliveryConfig)
+          }
+            .succeeded()
+        }
+      }
+
+      context("a new artifact passes stateful constraints while older versions are pending") {
+        deriveFixture {
+          copy(environment = Environment(
+            name = "staging",
+            constraints = setOf(DependsOnConstraint("test"), ManualJudgementConstraint())
+          ))
+        }
+
+        before {
+          // TODO: sucks that this is necessary but when using deriveFixture you get a different mockk
```

Comment:
The docs say you can use `modifyFixture` with `parentFixture` to modify the original fixture, which should include the mocks if that's where they're defined.

Reply:
Nice!